File: ScalaPatterns/src/main/scala/behavioral/Value.scala
package behavioral

/**
 * Value object is a small immutable object that represents a simple entity whose equality isn’t based on identity. Value objects are equal if all their fields are equal.
 * Value objects are widely used to represent numbers, dates, colors, etc. In enterprise applications they are used as DTOs a for inter-process communication.
 * Because of the immutability, value objects are handy in multi-threaded programing.
 *
 * Case classes exports constructor parameters as properties. By default, case classes are immutable.
 * Another approach could be using tuple sets
 *
 * @author Daniel Leon
 */
object Value {

  case class Point(x: Int, y: Int)

  def main(args:Array[String]) {
    //first method using case classes
    val firstPoint = Point(1, 2)
    val secondPoint = Point(1, 3)

    firstPoint match {
      case matchPoint if matchPoint.x == secondPoint.x && matchPoint.y == secondPoint.y => println("First approach : The points are the same")
      case _ => println("First approach : The points are different")
    }

    //second method using tuple
    val tuple = (1,2)
    tuple match {
      case (1,3) => println("Second approach : The points are the same")
      case _ => println("Second approach : The points are different")
    }
  }
}

File: JavaPatterns/src/main/java/behavioral/Value.java
package behavioral;

/**
 * Value object is a small immutable object that represents a simple entity whose equality isn’t based on identity. Value objects are equal if all their fields are equal.
 Value objects are widely used to represent numbers, dates, colors, etc. In enterprise applications they are used as DTOs a for inter-process communication. Because of the immutability,
 value objects are handy in multi-threaded programing.
 In Java, there’s no special syntax for value objects, so we have to define (and update) a constructor, getters and auxiliary methods explicitly
 */

class Point {
    private final int x, y;

    public Point(int x, int y) { this.x = x; this.y = y; }

    public int getX() { return x; }

    public int getY() { return y; }

    public boolean equals(Object o) {
        Point that = (Point) o;
        return x == that.x && y == that.y;
    }

    public int hashCode() {
        return 31 * x + y;
    }

    public String toString() {
        return String.format("Point(%d, %d)", x, y);
    }
}

public class Value {

    public static void main(String[] args) {

        Point point = new Point(1, 2);
        Point point1 = new Point(1, 2);
        System.out.println(point.equals(point1) ? "The points are the same" : "The points are different");
    }

}
